N-4

varfik opened the pull request.

varfik committed:
Update natural.cc ADD_NN_N

varfik committed:
Update natural.cc ADD_NN_N

uint8_t replaced Digit

eggunof commented on libraries/numbers/src/natural.cc:
```diff
@@ -78,7 +78,46 @@ Natural Natural::operator/(const Natural &rhs) const { return {}; }
+  if (rhs.digits_.size() > this->digits_.size()) {
+    for (int i = 0; i <= (rhs.digits_.size() - this->digits_.size()); i++) {
+      this->digits_.insert(this->digits_.begin(), 0);
+    }
```
можно расширить сразу на нужное количество элементов 
```c++
digits_.insert(digits_.begin(), rhs.digits_.size() - digits_.size(), 0);
```

eggunof commented on libraries/numbers/src/natural.cc:
```diff
@@ -78,7 +78,46 @@ Natural Natural::operator/(const Natural &rhs) const { return {}; }
+    // обновляем перенос путём деления нацело суммы разрядов операндов и
+    // переноса
+    surplus = (old_digit + rhs_digit + surplus) / 10;
+  }
```
Лучше использовать итераторы digits_.rbegin() и digits_.rend():
```с++
auto lhs_it = digits_.rbegin();
auto rhs_it = rhs.digits_.rbegin();
```
Пройтись сначала по правому операнду:
```c++
while (rhs_it != rhs.digits_.rend()) {
    *lhs_it = ...;
    ++lhs_it;
    ++rhs_it;
  }
  ```
  Потом по тому, что осталось от левого:
  ```c++
while (lhs_it != digits_.rend() && surplus != 0) {
    *lhs_it = ...;
    ++lhs_it;
  }
```

eggunof commented:
Убрать `this->`, добавить тесты

eggunof commented:
Нет реализации
```c++
Natural Natural::operator+(const Natural &rhs) const
```

varfik committed:
Update natural.cc ADD_NN_N

removed this->; added reverse iterators; created Natural Natural::operator+(const Natural &rhs) const

varfik committed:
Update test_natural.cc ADD_NN_N

eggunof commented on libraries/numbers/src/natural.cc:
```diff
@@ -66,7 +66,53 @@ Natural &Natural::operator++() { return *this; }
+  if (rhs.digits_.size() > result.digits_.size()) {
+    result.digits_.insert(result.digits_.begin(),
+                          rhs.digits_.size() - result.digits_.size(), 0);
+  }
+  // Перенос в следующий разряд
+  Digit surplus = 0;
+  // Изначальное значение разряда левого операнда,
+  // значение разряда правого операнда
+  Digit old_digit, rhs_digit;
+  // Реверсные итераторы для операндов (справо налево)
+  auto lhs_it = result.digits_.rbegin();
+  auto rhs_it = rhs.digits_.rbegin();
+  // Обрабатываем разряды, где оба числа имеют значение
+  while (rhs_it != rhs.digits_.rend()) {
+    old_digit = *lhs_it;
+    rhs_digit = *rhs_it;
+    // Складываем текущие разряды операндов и перенос;
+    // Заменяем текущий разряд левого операнда на остаток от деления на 10
+    *lhs_it = (old_digit + rhs_digit + surplus) % 10;
+    // Обновляем перенос путём деления нацело суммы разрядов операндов и
+    // переноса
+    surplus = (old_digit + rhs_digit + surplus) / 10;
+    ++lhs_it;
+    ++rhs_it;
+  }
+  // Обрабатываем оставшиеся разряды текущего числа, если есть перенос
+  while (lhs_it != result.digits_.rend() && surplus != 0) {
+    old_digit = *lhs_it;
+    // Обновляем текущий разряд с учётом переноса
+    *lhs_it = (old_digit + surplus) % 10;
+    // Обновляем перенос
+    surplus = (old_digit + surplus) / 10;
+    ++lhs_it;
+  }
+  // Если избыток не равен 0, дописываем его в начало
+  if (surplus != 0) {
+    result.digits_.insert(result.digits_.begin(), surplus);
+  }
+  return result;
+}
```
Реализовать через `operator+=`:
```c++
Natural result = *this;
result += rhs;
return result;
```

eggunof commented on libraries/numbers/tests/test_natural.cc:
```diff
@@ -6,4 +6,11 @@
+}
+
+TEST(NaturalTest, ADD_NN_N) {
+  Natutal a(10099);
```
Natutal -> Natural

eggunof commented on libraries/numbers/tests/test_natural.cc:
```diff
@@ -6,4 +6,11 @@
+}
+
+TEST(NaturalTest, ADD_NN_N) {
+  Natutal a(10099);
+  Natural b(599999);
```
передавать числа в конструктор, как строку

eggunof reviewed (changes requested):
Пофиксить тест, тестировать `operator+`, больше тестов с реально длинными числами

varfik committed:
Update natural.cc ADD_NN_N

duplicates removed

varfik committed:
Update test_natural.cc ADD_NN_N

trying to run test

varfik committed:
Update test_natural.cc ADD_NN_N

fixed test

eggunof commented on libraries/numbers/tests/test_natural.cc:
```diff
@@ -6,4 +6,11 @@
+}
+
+TEST(NaturalTest, ADD_NN_N) {
+  Natutal a("10099");
```
Natutal

varfik committed:
Update test_natural.cc ADD_NN_N

natutal fixed

eggunof committed:
Merge branch 'natural' into ADD_NN_N

eggunof committed:
Merge branch 'natural' into ADD_NN_N

# Conflicts:
#	libraries/numbers/tests/test_natural.cc

eggunof committed:
Merge branch 'natural' into ADD_NN_N

eggunof committed:
formating and tests

eggunof committed:
Merge branch 'natural' into ADD_NN_N
